h/formatters/formatter.py

```diff
@@ -0,0 +1,19 @@
+from h.traversal import AnnotationContext
+
+
+class AnnotationFormatter:
+    """Interface for annotation formatters."""
+
+    def preload(self, annotation_ids):
+        """
+        Preload data to speed up rendering of large batches of annotations.
+
+        :param annotation_ids: List of annotation ids to load
+        """
+
+    def format(self, annotation_context: AnnotationContext):
+        """
+        Render a dict of data to merge into the main presentation data.
+
+        :return: A dict of data
+        """
```

Comment:
The previous version had an absolute essay here, but basically it boils down to:

 * Preload useful things to speed it up
 * Render a partial dict which is merged up stream

Reply:
I think the common thing to do when you wanted to implement an abstract base class (back in the days before actual [`ABC`'s](https://docs.python.org/3/library/abc.html#module-abc)) was for the base class's methods to `raise NotImplementedError` to signal that subclasses have to implement them. In modern Python you would write an `ABC` instead.

As a stylistic preference I'd rather avoid using inheritance with concrete base classes whenever possible. I think inheritance tends to make for code that's easy to write but hard to read, inflexible designs, tight coupling, prevents isolated unit testing, etc. I'd just rather avoid peppering even harmless cases of it around when not necessary as I don't want to encourage it. And in this case `AnnotationFormatter` isn't really trying to be a concrete base class (the methods have no bodies). It's trying to be an interface or protocol. But technically it's actually a concrete base class.

Alternatives:

* Just don't bother with either an interface or a base class. Just let it be an "informal protocol" (a bunch of classes happen to implement the same interface but that interface isn't separately documented anywhere like in a base class or formal interface). This is honestly very likely the best solution.

* Or perhaps define the protocol by documentation only in a `formatters/README.md` file (that no one will ever read or keep up to date :man_shrugging:)

* Make it an ABC (subclass `abc.ABC`). This is the Right Way to write a class that's designed to be an interface in modern Python.

  A nice thing about ABC's is that you can [register](https://docs.python.org/3/library/abc.html#abc.ABCMeta.register) subclasses of them without using inheritance:

  ```python
  from h.formatters import AnnotationFormatter

  @AnnotationFormatter.register
  class SomeFormatter:
      ...
  ```

* I've never used them but Python 3.8 added "protocol classes" to Python: <https://www.python.org/dev/peps/pep-0544/>, <https://docs.python.org/3/library/typing.html#typing.Protocol> so that you can do "structural subtyping" a.k.a. "static duck-typing". But I'm not sure if these are really relevant / necessary here.

Reply:
I should give the new version a whirl at some point. I always found the old ABCs a bit clunky. In this case I think I'll just get rid of it. I'm hoping to kill off these classes anyway.